paywall/src/unlock.js/Wallet.ts

```diff
@@ -0,0 +1,315 @@
+ * user accounts iframe and initialize it
+ */
+export default class Wallet {
+  private readonly iframes: IframeHandler
+  private readonly window: Web3Window
+  private readonly hasWallet: boolean = true
+  private readonly isMetamask: boolean
+  private readonly config: PaywallConfig
+  private hasWeb3: boolean = false
+  private useUserAccounts: boolean = false
+
+  private userAccountAddress: string | null = null
+  private userAccountNetwork: number
+
+  constructor(
+    window: Web3Window,
+    iframes: IframeHandler,
+    config: PaywallConfig,
+    constants: StartupConstants
+  ) {
+    this.window = window
+    this.iframes = iframes
+    this.config = config
+    this.userAccountNetwork = constants.network
+
+    // do we have a web3 wallet?
+    this.hasWallet = !!(this.window.web3 && this.window.web3.currentProvider)
+    this.isMetamask = !!(
+      window.web3 &&
+      window.web3.currentProvider &&
+      window.web3.currentProvider.isMetamask
+    )
+    // user accounts are used in 2 conditions:
+    // 1. there is no crypto wallet present
+    // 2. the paywall configuration explicitly asks for them
+    this.useUserAccounts =
+      !this.hasWallet &&
+      (config.unlockUserAccounts === true ||
+        config.unlockUserAccounts === 'true')
+    if (process.env.DEBUG) {
+      if (this.useUserAccounts) {
+        // eslint-disable-next-line
+        console.log('[USER ACCOUNTS] using user accounts')
+      } else {
+        // eslint-disable-next-line
+        console.log('[USER ACCOUNTS] using native crypto wallet')
+      }
+    }
+  }
+
+  init() {
+    if (this.useUserAccounts) {
+      // create the preconditions for using user accounts
+      this.setupUserAccounts()
+    }
+    // set up the proxy wallet
+    this.setupProxyWallet()
+  }
+
+  /**
+   * This is called only if we can use user accounts
+   */
+  setupUserAccounts() {
+    // listen for updates to state from the data iframe, and forward them to the checkout UI
+    this.iframes.data.on(PostMessages.UPDATE_LOCKS, locks =>
+      this.iframes.accounts.postMessage(PostMessages.UPDATE_LOCKS, locks)
+    )
+
+    // pass on the configuration and request the latest data
+    this.iframes.accounts.on(PostMessages.READY, () => {
+      this.iframes.accounts.postMessage(PostMessages.CONFIG, this.config)
+
+      this.iframes.data.postMessage(PostMessages.SEND_UPDATES, 'locks')
+    })
+
+    // listen for account and network from the user accounts iframe
+    this.iframes.accounts.on(PostMessages.UPDATE_ACCOUNT, account => {
+      this.userAccountAddress = account
+    })
+    this.iframes.accounts.on(PostMessages.UPDATE_NETWORK, network => {
+      this.userAccountNetwork = network
+    })
+
+    // when a purchase is in progress, tell the data iframe to retrieve the transaction
+    this.iframes.accounts.on(PostMessages.INITIATED_TRANSACTION, () => {
+      this.iframes.data.postMessage(
+        PostMessages.INITIATED_TRANSACTION,
+        undefined
```

Comment:
Do you mind putting a comment about the `undedined`? like what should it be?

Reply:
this is a limitation of the typing. The postMessage does not actually need the parameter, but the current version of types does not allow the 2nd parameter to be missing. I have a branch I made a month ago that fixes this so that we don't have to pass the `undefined`, but I considered it lower priority than actually making it work. Once all of this stuff is merged, I planned to submit that branch as a PR, so the code is more natural.

Reply:
Sorry for the confusuin here. I was hoping you would add  a comment indicated what that value should be.
`undefined /* thething */` ;)
Thanks!